integrationtests/src/main/java/io/kroxylicious/proxy/testcluster/ContainerBasedKafkaCluster.java

```diff
@@ -0,0 +1,226 @@
+            }
+        };
+        Supplier<Endpoint> zookeeperEndpointSupplier = () -> new Endpoint("zookeeper", ContainerBasedKafkaCluster.ZOOKEEPER_PORT);
+        this.brokers = clusterConfig.getBrokerConfigs(endPointConfigSupplier, zookeeperEndpointSupplier).map(holder -> {
+            String netAlias = "broker-" + holder.getBrokerNum();
+            MountableFile kafkaJaasConf = MountableFile.forClasspathResource("kafka_jaas.conf");
+            KafkaContainer kafkaContainer = new KafkaContainer(this.kafkaImage)
+                    .withNetwork(this.network)
+                    .withNetworkAliases(netAlias)
+                    .withEnv("KAFKA_BROKER_ID", holder.getBrokerNum() + "")
+                    .withEnv("ALLOW_PLAINTEXT_LISTENER", "yes")
+                    .withEnv("BITNAMI_DEBUG", "true")
+                    .withEnv("XSHELLOPTS", "xtrace")
+                    .withStartupTimeout(Duration.ofMinutes(2))
+                    .withCopyFileToContainer(kafkaJaasConf, "/opt/bitnami/kafka/config/");
+
+            kafkaContainer.addFixedExposedPort(holder.getExternalPort(), KAFKA_PORT);
+
+            // Once https://github.com/docker-java/docker-java/pull/1963 is available, we will be able to copy
+            // these properties to the container as server.properties, and avoid the need to map the properties
+            // into environment variables.
+            // kafkaContainer.withCopyFileToContainer(..properties.., "/opt/bitnami/kafka/config/");
+
+            kafkaContainer.withEnv(serverPropertiesToBitnamiEnvVar(holder.getProperties()));
+
+            if (this.clusterConfig.isKraftMode()) {
+                return kafkaContainer
+                        .withEnv("KAFKA_ENABLE_KRAFT", "yes")
+                        .withEnv("KAFKA_KRAFT_CLUSTER_ID", holder.getKafkaKraftClusterId());
+            }
+            else {
+                return kafkaContainer.dependsOn(this.zookeeper);
+            }
+        }).collect(Collectors.toList());
+    }
+
+    private Map<String, String> serverPropertiesToBitnamiEnvVar(Properties serverProperties) {
+        Map<String, String> envVars = new HashMap<>();
+        serverProperties.entrySet().forEach(p -> envVars.put("KAFKA_CFG_" + p.getKey().toString().toUpperCase().replace('.', '_'), p.getValue().toString()));
+        return envVars;
+    }
+
+    @Override
+    public String getBootstrapServers() {
+        return brokers.stream()
+                .map(b -> String.format("localhost:%d", b.getMappedPort(KAFKA_PORT)))
+                .collect(Collectors.joining(","));
+    }
+
+    private Stream<GenericContainer<?>> allContainers() {
+        return Stream.concat(
+                this.brokers.stream(),
+                Stream.ofNullable(this.zookeeper));
+    }
+
+    @Override
+    @SneakyThrows
+    public void start() {
+        try {
+            if (zookeeper != null) {
```

Comment:
Do we need to start zookeeper if we are in KRaft mode?

Reply:
In KRaft mode, the `zookeeper` will be set to null, so it won't be initialized

Reply:
Ahh thanks @showuon